File: iomerge-server-app/src/main/java/com/github/krzychek/iomerge/server/utils/ChainHelper.java
package com.github.krzychek.iomerge.server.utils;

import com.github.krzychek.iomerge.server.api.Chainable;

import java.lang.reflect.Proxy;
import java.util.List;


public class ChainHelper {

	@SuppressWarnings("unchecked")
	private static <T> T getNOOPChainLink(Class<T> clazz) {
		return (T) Proxy.newProxyInstance(clazz.getClassLoader(), new Class[]{clazz}, (proxy, method, args) -> null);
	}

	public static <T extends Chainable<T>> T createChain(List<T> chain, Class<T> clazz) {
		//noinspection OptionalGetWithoutIsPresent
		chain.stream()
				.reduce((listener1, listener2) -> {
					listener1.chain(listener2);
					return listener2;
				})
				.get()
				.chain(getNOOPChainLink(clazz));

		return chain.get(0);
	}
}

File: iomerge-server-app/src/main/java/com/github/krzychek/iomerge/server/utils/ChainHelper.kt
package com.github.krzychek.iomerge.server.utils

import com.github.krzychek.iomerge.server.api.Chainable

import java.lang.reflect.Proxy


object ChainHelper {

	@Suppress("UNCHECKED_CAST")
	private fun <T> getNOOPChainLink(clazz: Class<T>): T {
		return Proxy.newProxyInstance(clazz.classLoader, arrayOf(clazz), { proxy, method, args -> null }) as T
	}

	fun <T : Chainable<T>> createChain(chain: List<T>, clazz: Class<T>): T {
		return chain.foldRight(getNOOPChainLink(clazz),
				{ listener1, listener2 ->
					listener1.chain(listener2)
					listener1
				})
	}
}
